src/maxblocksize.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2009-2010 Satoshi Nakamoto
+// Copyright (c) 2009-2017 The Bitcoin developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "maxblocksize.h"
+
+#include "chain.h"
+#include "util.h"
+#include <boost/lexical_cast.hpp>
+#include <string>
+
+uint64_t GetNextMaxBlockSize(const CBlockIndex *pindexLast, const Consensus::Params &params)
+{
+    // BIP100 not active, return legacy max size
+    if (pindexLast == NULL || pindexLast->nHeight < params.bip100ActivationHeight)
+        return BLOCKSTREAM_CORE_MAX_BLOCK_SIZE;
+
+    uint64_t nMaxBlockSize = pindexLast->nMaxBlockSize;
+
+    // Only change once per difficulty adjustment interval
+    if ((pindexLast->nHeight + 1) % params.DifficultyAdjustmentInterval() != 0)
+    {
+        return nMaxBlockSize;
+    }
+
+    std::vector<uint64_t> votes;
+    const CBlockIndex *pindexWalk = pindexLast;
+    for (int64_t i = 0; i < params.DifficultyAdjustmentInterval(); i++)
+    {
+        assert(pindexWalk);
+        assert(pindexWalk->nMaxBlockSize == nMaxBlockSize);
+        votes.push_back(pindexWalk->nMaxBlockSizeVote ? pindexWalk->nMaxBlockSizeVote : pindexWalk->nMaxBlockSize);
+        pindexWalk = pindexWalk->pprev;
+    }
+
+    std::sort(votes.begin(), votes.end());
+    uint64_t lowerValue = votes.at(params.nMaxBlockSizeChangePosition - 1);
+    uint64_t raiseValue = votes.at(params.DifficultyAdjustmentInterval() - params.nMaxBlockSizeChangePosition);
+
+    assert(lowerValue >= 1000000); // minimal vote supported is 1MB
```

Comment:
this is a bit dangerous, since the FindVote accepts 0.  Regardless of whether the FindVote gets fixed, I'd prefer a forcing function here not an assert, because BIP100 makes is clear that 1MB is the lowest you can go and so if somehow other portions of the code get "faked out" we can fall back on that value rather than crashing:
if (lowerValue < 1000000) lowerValue = 1000000;